lib/mrtrix3/app.py

```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
            not group == self._positionals and \
            group.title not in ( 'options', 'optional arguments' )
 
+  class TypeBoolean:
+    def __call__(self, input_value):
+      true_var = "True"
+      false_var = "False"
+      converted_value = None
+      if input_value.lower() == true_var.lower():
+        converted_value = True
+      elif input_value.lower() == false_var.lower():
+        converted_value = False
+      else:
+        raise argparse.ArgumentTypeError('Entered value is not of type boolean')
+      return converted_value
+
+  class TypeIntegerSequence:
+    def __call__(self, input_value):
+      seq_elements = input_value.split(',')
+      int_list = []
+      for i in seq_elements:
+        try:
+          converted_value = int(i)
+          int_list.append(converted_value)
+        except:
+          raise argparse.ArgumentTypeError('Entered value is not an integer sequence')
+      return int_list
+  
+  class TypeFloatSequence:
+    def __call__(self, input_value):
+      seq_elements = input_value.split(',')
+      float_list = []
+      for i in seq_elements:
+        try:
+          converted_value = float(i)
+          float_list.append(converted_value)
+        except:
+          argparse.ArgumentTypeError('Entered value is not a float sequence')
+      return float_list
+
+  class TypeInputDirectory:
+    def __call__(self, input_value):
+      if not os.path.exists(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' does not exist')
+      elif not os.path.isdir(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' is not a directory')
+      else:
+        return input_value
+
+  class TypeOutputDirectory:
+    def __call__(self, input_value):
+      return input_value
+
+  class TypeInputFile:
+    def __call__(self, input_value):
+      if not os.path.exists(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' path does not exist')
+      elif not os.path.isfile(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' is not a file')
+      else:
+        return input_value
+
+  class TypeOutputFile:
+    def __call__(self, input_value):
+      return input_value
+
+  class TypeInputImage:
+    def __call__(self, input_value):
+      return input_value
+
+  class TypeOutputImage:
+    def __call__(self, input_value):
+      return input_value
+
+  class TypeInputTractogram:
+    def __call__(self, input_value):
+      if not input_value.endsWith('.tck'):
+        raise argparse.ArgumentTypeError(input_value + ' is not a valid track file')
+      return input_value 
```

Comment:
Tractograms may be an opportunity to use some inheritance. Any checks that apply to input / output files additionally apply to tractograms, but then checking the file extension is an *additional* task that applies to tractogram files but not generic files.

Reply:
Changes available in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/8d91a38c27bb2340584b4c3ecb5d3c9556dd6096

Reply:
That commit duplicates within `TypeInputTractogram` the checks previously performed only in `TypeInputFile`. However I was indicating that that should be achieved through class inheritance.

Reply:
Changes made in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/6dcfe5353c171a7a86aa2109b175f667aecf1fb2